HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +61,79 @@ cpdef move_data_serial(np.int32_t[:] counts,
+                    counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+    cdef:
+        int idx, r
+        int first_to_lock, second_to_lock  # to implement ordered locking
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to
+    # parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+
+    with nogil:
+
+        # with gil:
+        #    print "entering fine grained"
+       
+        for r in range(repeat):
+            for idx in prange(src.shape[0], nogil=False, num_threads=4):
+
+                #####
+                #
+                # Lock-ordering
+
+                # We define a total ordering on the set of objects eligible
+                # for locking and use this ordering to choose the sequence of
+                # lock acquisition. In our case, src[idx] and dest[idx]
+                # represent elements of our counts array; we always lock a
+                # lower index of the counts array before a higher index of the
+                # counts array, otherwise we may encounter deadlock.
+                #
+                # Note we also add a check to ensure that we do not lock (or
+                # do anything) if the src and dest indices are the same.
+
+                if src[idx]!=dest[idx]:
+
+                    if src[idx]<dest[idx]:
+                        first_to_lock = src[idx]
+                        second_to_lock = dest[idx]
+                    else:
+                        first_to_lock = dest[idx]
+                        second_to_lock = src[idx]
+                
+                    omp_set_lock(&(locks[first_to_lock]))
+                    omp_set_lock(&(locks[second_to_lock]))
+               
+                    ### critical section, increment and decrement counters
+                    if counts[src[idx]] > 0:
+                        counts[src[idx]] -= 1
+                        counts[dest[idx]] += 1
+                    ##### #####
+
+                    # remove locks in reverse order
+                    omp_unset_lock(&(locks[second_to_lock]))
+                    omp_unset_lock(&(locks[first_to_lock]))
```

Comment:
Looks good! They provided code `acquire` and `release` helper functions, but `omp_set_lock` and `omp_unset_lock` work just the same.